src/docs/wasm-compilation-pipeline.md

```diff
@@ -7,19 +7,23 @@ WebAssembly is a binary format that allows you to run code from programming lang
+This is why _hot_ functions, which are functions that get executed often, get re-compiled with [TurboFan](/docs/turbofan), the optimizing compiler in V8 for both WebAssembly and JavaScript. TurboFan is a [multi-pass compiler](https://en.wikipedia.org/wiki/Multi-pass_compiler), which means that it builds multiple internal representations of the compiled code before emitting machine code. These additional internal representations allow optimizations and better register allocations, resulting in significantly faster code.
 
-TurboFan compiles the WebAssembly module function by function. As soon as one function finishes, it immediately replaces the function compiled by Liftoff. Any new calls to that function will then use the new, optimized code produced by TurboFan, not the Liftoff code. Note though that we don’t do on-stack-replacement. This means that if TurboFan finishes optimizing a function that was already invoked when only the Liftoff version was available, it will finish its execution using the Liftoff version.
+V8 monitors how often WebAssembly functions get called. Once a function reaches a certain threshold, the function is considered _hot_, and re-compilation gets triggered on a background thread. Once compilation is finished, the new code gets registered with the WebAssembly module, replacing the existing Liftoff code. Any new calls to that function will then use the new, optimized code produced by TurboFan, not the Liftoff code. Note though that we don’t do on-stack-replacement. This means that if TurboFan code gets available after the function gets called, the function call will complete its execution with Liftoff code.
 
 ## Code caching
 
-If the WebAssembly module was compiled with `WebAssembly.compileStreaming`, then the TurboFan-generated machine code will also get cached. When the same WebAssembly module is fetched again from the same URL, the module is not compiled but loaded from cache. More information about code caching is available [in a separate blog post](/blog/wasm-code-caching).
+If the WebAssembly module was compiled with `WebAssembly.compileStreaming`, then the TurboFan-generated machine code will also get cached. When the same WebAssembly module is fetched again from the same URL, then the cached code can be used immediately, without additional compilation. More information about code caching is available [in a separate blog post](/blog/wasm-code-caching).
+
+Code caching gets triggered whenever the amount of generated TurboFan code reaches a certain threshold. This means that for large WebAssembly modules the TurboFan code gets cached incrementally, whereas for small WebAssembly modules the TurboFan code may never get cached. Liftoff code does not get cached, as Liftoff is nearly as fast compiling code as code is loaded from the cache.
```

Comment:
Maybe "... Liftoff compilation is nearly as fast as loading code from the cache."?

Reply:
Done.